File: stripe_terminal/android/src/main/kotlin/mek/stripeterminal/mappings/SimulatorConfigurationMappings.kt
package mek.stripeterminal.mappings

import com.stripe.stripeterminal.external.models.SimulateReaderUpdate
import com.stripe.stripeterminal.external.models.SimulatedCard
import com.stripe.stripeterminal.external.models.SimulatedCardType
import com.stripe.stripeterminal.external.models.SimulatorConfiguration
import mek.stripeterminal.api.SimulateReaderUpdateApi
import mek.stripeterminal.api.SimulatedCardApi
import mek.stripeterminal.api.SimulatedCardTypeApi
import mek.stripeterminal.api.SimulatorConfigurationApi

fun SimulatorConfigurationApi.toHost(): SimulatorConfiguration {
    return SimulatorConfiguration(
        update = update.toHost(),
        simulatedCard = simulatedCard.toHost(),
        simulatedTipAmount = simulatedTipAmount,
    )
}

fun SimulateReaderUpdateApi.toHost(): SimulateReaderUpdate {
    return when (this) {
        SimulateReaderUpdateApi.AVAILABLE -> SimulateReaderUpdate.UPDATE_AVAILABLE
        SimulateReaderUpdateApi.NONE -> SimulateReaderUpdate.NONE
        SimulateReaderUpdateApi.REQUIRED -> SimulateReaderUpdate.REQUIRED
        SimulateReaderUpdateApi.RANDOM -> SimulateReaderUpdate.RANDOM
    }
}

fun SimulatedCardApi.toHost(): SimulatedCard {
    if (type != null) return SimulatedCard(type.toHost())
    return SimulatedCard(testCardNumber!!)
}

fun SimulatedCardTypeApi.toHost(): SimulatedCardType {
    return when (this) {
        SimulatedCardTypeApi.VISA -> SimulatedCardType.VISA
        SimulatedCardTypeApi.VISA_DEBIT -> SimulatedCardType.VISA_DEBIT
        SimulatedCardTypeApi.VISA_US_COMMON_DEBIT -> SimulatedCardType.VISA_US_COMMON_DEBIT
        SimulatedCardTypeApi.MASTERCARD -> SimulatedCardType.MASTERCARD
        SimulatedCardTypeApi.MASTER_DEBIT -> SimulatedCardType.MASTERCARD_DEBIT
        SimulatedCardTypeApi.MASTERCARD_PREPAID -> SimulatedCardType.MASTERCARD_PREPAID
        SimulatedCardTypeApi.AMEX -> SimulatedCardType.AMEX
        SimulatedCardTypeApi.AMEX2 -> SimulatedCardType.AMEX_2
        SimulatedCardTypeApi.DISCOVER -> SimulatedCardType.DISCOVER
        SimulatedCardTypeApi.DISCOVER2 -> SimulatedCardType.DISCOVER_2
        SimulatedCardTypeApi.DINERS -> SimulatedCardType.DINERS
        SimulatedCardTypeApi.DINERS14_DIGIT -> SimulatedCardType.DINERS_14_DIGITS
        SimulatedCardTypeApi.JBC -> SimulatedCardType.JCB
        SimulatedCardTypeApi.UNION_PAY -> SimulatedCardType.UNION_PAY
        SimulatedCardTypeApi.INTERAC -> SimulatedCardType.INTERAC
        SimulatedCardTypeApi.EFTPOS_AU_DEBIT -> SimulatedCardType.EFTPOS_AU_DEBIT
        SimulatedCardTypeApi.EFTPOS_AU_VISA_DEBIT -> SimulatedCardType.EFTPOS_AU_VISA_DEBIT
        SimulatedCardTypeApi.EFTPOS_AU_DEBIT_MASTERCARD -> SimulatedCardType.EFTPOS_AU_DEBIT_MASTERCARD
        SimulatedCardTypeApi.CHARGE_DECLINED -> SimulatedCardType.CHARGE_DECLINED
        SimulatedCardTypeApi.CHARGE_DECLINED_INSUFFICIENT_FUNDS -> SimulatedCardType.CHARGE_DECLINED_INSUFFICIENT_FUNDS
        SimulatedCardTypeApi.CHARGE_DECLINED_LOST_CARD -> SimulatedCardType.CHARGE_DECLINED_LOST_CARD
        SimulatedCardTypeApi.CHARGE_DECLINED_STOLEN_CARD -> SimulatedCardType.CHARGE_DECLINED_STOLEN_CARD
        SimulatedCardTypeApi.CHARGE_DECLINED_EXPIRED_CARD -> SimulatedCardType.CHARGE_DECLINED_EXPIRED_CARD
        SimulatedCardTypeApi.CHARGE_DECLINED_PROCESSING_ERROR -> SimulatedCardType.CHARGE_DECLINED_PROCESSING_ERROR
        SimulatedCardTypeApi.ONLINE_PIN_CVM -> SimulatedCardType.ONLINE_PIN_CVM
        SimulatedCardTypeApi.ONLINE_PIN_SCA_RETRY -> SimulatedCardType.ONLINE_PIN_SCA_RETRY
        SimulatedCardTypeApi.OFFLINE_PIN_CVM -> SimulatedCardType.OFFLINE_PIN_CVM
        SimulatedCardTypeApi.OFFLINE_PIN_SCA_RETRY -> SimulatedCardType.OFFLINE_PIN_SCA_RETRY
    }
}

File: stripe_terminal/ios/Classes/Mappings/SimulatorConfigurationMappings.swift
//
//  SimulatorConfigurationMappings.swift
//  mek_stripe_terminal
//
//  Created by Kuama on 25/11/23.
//

import Foundation
import StripeTerminal

extension SimulateReaderUpdateApi {
    func toHost() -> SimulateReaderUpdate {
        switch self {
        case SimulateReaderUpdateApi.available:
            return SimulateReaderUpdate.available
        case SimulateReaderUpdateApi.none:
            return SimulateReaderUpdate.none
        case SimulateReaderUpdateApi.required:
            return SimulateReaderUpdate.required
        case SimulateReaderUpdateApi.random:
            return SimulateReaderUpdate.random
        }
    }
}

extension SimulatedCardApi {
    func toHost() -> SimulatedCard {
        if let type {
            return SimulatedCard(type: type.toHost())
        }
        return SimulatedCard(testCardNumber: testCardNumber ?? "")
    }

}

extension SimulatedCardTypeApi {
    func toHost() -> SimulatedCardType {
        switch self {
        case SimulatedCardTypeApi.visa:
            return SimulatedCardType.visa
        case SimulatedCardTypeApi.visaDebit:
            return SimulatedCardType.visaDebit
        case SimulatedCardTypeApi.visaUsCommonDebit:
            return SimulatedCardType.visaUsCommonDebit
        case SimulatedCardTypeApi.mastercard:
            return SimulatedCardType.mastercard
        case SimulatedCardTypeApi.masterDebit:
            return SimulatedCardType.masterDebit
        case SimulatedCardTypeApi.mastercardPrepaid:
            return SimulatedCardType.mastercardPrepaid
        case SimulatedCardTypeApi.amex:
            return SimulatedCardType.amex
        case SimulatedCardTypeApi.amex2:
            return SimulatedCardType.amex2
        case SimulatedCardTypeApi.discover:
            return SimulatedCardType.discover
        case SimulatedCardTypeApi.discover2:
            return SimulatedCardType.discover2
        case SimulatedCardTypeApi.diners:
            return SimulatedCardType.diners
        case SimulatedCardTypeApi.diners14Digit:
            return SimulatedCardType.diners14Digit
        case SimulatedCardTypeApi.jbc:
            return SimulatedCardType.jcb
        case SimulatedCardTypeApi.unionPay:
            return SimulatedCardType.unionPay
        case SimulatedCardTypeApi.interac:
            return SimulatedCardType.interac
        case SimulatedCardTypeApi.eftposAuDebit:
            return SimulatedCardType.eftposAuDebit
        case SimulatedCardTypeApi.eftposAuVisaDebit:
            return SimulatedCardType.eftposAuDebit
        case SimulatedCardTypeApi.eftposAuDebitMastercard:
            return SimulatedCardType.eftposAuDebitMastercard
        case SimulatedCardTypeApi.chargeDeclined:
            return SimulatedCardType.chargeDeclined
        case SimulatedCardTypeApi.chargeDeclinedInsufficientFunds:
            return SimulatedCardType.chargeDeclinedInsufficientFunds
        case SimulatedCardTypeApi.chargeDeclinedLostCard:
            return SimulatedCardType.chargeDeclinedLostCard
        case SimulatedCardTypeApi.chargeDeclinedStolenCard:
            return SimulatedCardType.chargeDeclinedStolenCard
        case SimulatedCardTypeApi.chargeDeclinedExpiredCard:
            return SimulatedCardType.chargeDeclinedExpiredCard
        case SimulatedCardTypeApi.chargeDeclinedProcessingError:
            return SimulatedCardType.chargeDeclinedProcessingError
        case SimulatedCardTypeApi.onlinePinCvm:
            return SimulatedCardType.onlinePinCvm
        case SimulatedCardTypeApi.onlinePinScaRetry:
            return SimulatedCardType.onlinePinScaRetry
        case SimulatedCardTypeApi.offlinePinCvm:
            return SimulatedCardType.offlinePinCvm
        case SimulatedCardTypeApi.offlinePinScaRetry:
            return SimulatedCardType.offlinePinScaRetry
        }
    }
}
